src/com/fishrock123/entitysuppressor/configheader.txt
###############################################################

 EntitySuppressor v3 Configuration
 
     Developer: Fishrock123
          Report bugs and feature requests to: http://dev.bukkit.org/server-mods/entitysuppressor/tickets/
     Support:
          @Fishrock123 on Twitter
          Fishrock123 on www.bukkit.org or dev.bukkit.org
          http://forums.bukkit.org/threads/fix-admn-mech-entitysuppressor-v1-4-limit-monsters-reduce-lag-1597.39794/
          http://dev.bukkit.org/server-mods/entitysuppressor/forum/

 REMEMBER: Animals are Mobs too!

###############################################################


 Set what the plugin should limit.
 LimitMonsters: true / false
 LimitSquid: true / false
 LimitAnimals: true / false
 LimitNPCs: true / false (Alternate: LimitTESTIFICATES: true / false)
 LimitBats: true / false 


 Set the total maximum monsters per world.
 MonsterMaximums:
   world: ~ 32-64 per player
   world_nether: ~ 32-64 per player
   world_the_end: ~ 32-64 per player
  
 Set the total maximum squid per world.
 SquidMaximums:
   world: ~ 1-4 per player
   
 Set the total maximum animals per world.
 AnimalMaximums:
   world: ~ 12-24 per player
   
   
 The dynamic chunk-calculated maximums are calculated using this formula:
 (loaded_chunks * config_value) / 256
   
 Set the dynamically calculated maximum monsters per world.
 ChunkCalculatedMonsterMaximums:
   world: 23
   world_nether: 23
  
 Set the dynamically calculated maximum squid per world.
 ChunkCalculatedSquidMaximums:
   world: 6
   
 Set the dynamically calculated maximum animals per world.
 ChunkCalculatedAnimalMaximums:
   world: 23
   

 Set number of server ticks between checks to allow additional mobs.
 ScanInterval: 100

 Set how many entities' distances should be checked for removal per scan.
 EntitiesPerScan: 64
 
 
 Set whether to use dynamic monster removal and spawn prevention (Increases difficulty back to that of vanilla)
 UseMonsterDistanceRemoval: true / false
 
 Set the distance at which any mobs farther away from all players should be removed each scan.
 RemoveAtDistance: 128
 
 Set the distance at which monsters will have their spawn's canceled.
 CancelSpawnAtDistance: 92

 Set if monster's spawns should be canceled if their height is not close to the players' heights.
 OnlySpawnOnCloseLayers: true / false

 Set the maximum number of layers above or below that a monster can spawn on.
 MaximumSpawnLayerDeviation: 20


 Set the number of ticks between the times the server tries to spawn monsters.
 TicksPerMonsterSpawns: 10

 Set the number of ticks between the times the server tries to spawn animals.
 TicksPerAnimalSpawns: 800


 Set whether to keep the spawn chunks in memory.
 KeepSpawnChunksInMemory: true / false
 

 Get extra debugging information
 Debug: true / false

###############################################################

EntitySuppressor/src/Fishrock123/EntitySuppressor/configheader.txt
###############################################################

 EntitySuppressor v2 Configuration
 
     Developer: Fishrock123
          Report bugs and feature requests to: http://dev.bukkit.org/server-mods/entitysuppressor/tickets/
     Support:
          @Fishrock123 on Twitter
          Fishrock123 on www.bukkit.org or dev.bukkit.org
          http://forums.bukkit.org/threads/fix-admn-mech-entitysuppressor-v1-4-limit-monsters-reduce-lag-1597.39794/
          http://dev.bukkit.org/server-mods/entitysuppressor/forum/

 REMEMBER: Animals are Mobs too!
 
 ##############################################################
 
 In brackets following suggested values are general setting levels, this is a rough guide:
 
 Very Low: Under 2GB ram, Under 2Ghz processor speed. (~1-6 players)
 Low: 3 - 4GB ram, 2 - 2.3Ghz processor speed. (~6-12 players)
 Medium (mid): 4 - 6GB ram, 2.3 - 2.7Ghz processor speed. (~12-24 players)
 High: 6 - 8+ GB ram, 2.8+Ghz processor speed. (~24-32+ players)

###############################################################

 Set what the plugin should limit.
 LimitMonsters: true / false
 LimitSquid: true / false
 LimitAnimals: true / false
 

 Set the total maximum monsters per world.
 MonsterMaximums:
   world: 128 (very low) / 256 (low) / 512 (mid) / 1024 (high)
   world_nether: 128 (very low) / 256 (low) / 512 (mid) / 1024 (high)
   world_the_end: 256 (no less than this)
  
 Set the total maximum squid per world.
 SquidMaximums:
   world: 16 (very low) / 32 (low) / 64 (mid) / 128 (high)
   
 Set the total maximum animals per world.
 AnimalMaximums:
   world: 64 (very low) / 128 (low) 256 (mid) / 512 (high)
   
   
 The dynamic chunk-calculated maximums are calculated using this formula:
 ( ( Number of loaded chunks ) times ( config value ) ) divided by ( 256 )
 (loaded_chunks * X) / 256
   
 Set the dynamically calculated maximum monsters per world.
 ChunkCalculatedMonsterMaximums:
   world: 23
   world_nether: 23
  
 Set the dynamically calculated maximum squid per world.
 ChunkCalculatedSquidMaximums:
   world: 6
   
 Set the dynamically calculated maximum animals per world.
 ChunkCalculatedAnimalMaximums:
   world: 23
   

 Set number of server ticks between checks to allow additional mobs.
 ScanInterval: 100
 
 
 Set whether to use dynamic monster removal and spawn prevention (Increases difficulty back to that of vanilla)
 UseMonsterDistanceRemoval: true / false
 
 Set the distance at which any mobs farther away from all players should be removed each scan.
 RemoveAtDistance: 128
 
 Set the distance at which monsters will have their spawn's cancelled.
 CancelSpawnAtDistance: 92
 

 Get extra debugging information
 Debug: true / false

#######   OUTDATED SETTINGS - REMOVE IF YOU HAVE THEM #######

    ESConfigVersion
        Removed.
        
    checkDifference
        Replaced by a square root function.

    maxMonsters / DefaultMaximum
        Removed, everything must now be set per world.

    nonLimitedWorlds:
        Removed, set that world's maximum to 9001.
        
    SpawnFlagsCheckInterval:
        Replaced by ScanInterval.

###############################################################

src/com/fishrock123/entitysuppressor/Config.java
package com.fishrock123.entitysuppressor;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map.Entry;

import org.bukkit.configuration.file.FileConfiguration;

import com.fishrock123.entitysuppressor.utils.Utils;

public class Config {
	public static int interval;
	public static boolean lMonsters;
	public static boolean lSquid;
	public static boolean lAnimals;
	public static boolean lBats;
	public static boolean lNPCs;
	public static boolean d;
	public static boolean lSpawners;
	public static boolean uRemoveM;
	public static boolean uLRemove;
	public static int maxLDev;
	public static int sqRemovalDist;
	public static int sqCancelDist;
	public static int entitiesPerScan;
	public static int ticksPerMonsterSpawns;
	public static int ticksPerAnimalSpawns;
	public static boolean qSpawner;
	public static boolean keepSpawnInMem;
	public static boolean KSIM;
	
	private ESPlugin m;
	
	public Config(ESPlugin instance) {
		m = instance;
	}
	
	public void generate() {
		final FileConfiguration c = m.getConfig();
		if (!new File(m.getDataFolder(), "config.yml").exists()) {
			m.log("ES: Generating New Config File... ");
			c.addDefault("LimitMonsters", true);
			c.addDefault("LimitSquid", true);
			c.addDefault("LimitAnimals", true);
			c.addDefault("LimitBats", true);
			c.addDefault("LimitNPCs", true);
   	 		c.createSection("MonsterMaximums").addDefault("world", 256);
   	 		c.getConfigurationSection("MonsterMaximums").addDefault("world_nether", 128);
   	 		c.getConfigurationSection("MonsterMaximums").addDefault("world_the_end", 256);
   	 		c.createSection("ChunkCalculatedMonsterMaximums").addDefault("world", 23);
	 		c.getConfigurationSection("ChunkCalculatedMonsterMaximums").addDefault("world_nether", 23);
	 		c.createSection("SquidMaximums").addDefault("world", 32);
   	 		c.createSection("ChunkCalculatedSquidMaximums").addDefault("world", 6);
	 		c.createSection("AnimalMaximums").addDefault("world", 128);
   	 		c.createSection("ChunkCalculatedAnimalMaximums").addDefault("world", 23);
   	 		c.addDefault("ScanInterval", 100);
   	 		c.addDefault("EntitiesPerScan", 64);
   	 		c.addDefault("UseMonsterDistanceRemoval", true);
   	 		c.addDefault("RemoveAtDistance", 82);
   	 		c.addDefault("CancelSpawnAtDistance", 64);
   	 		c.addDefault("OnlySpawnOnCloseLayers", true);
   	 		c.addDefault("MaximumSpawnLayerDeviation", 20);
   	 		c.addDefault("TicksPerMonsterSpawns", 10);
   	 		c.addDefault("TicksPerAnimalSpawns", 800);
   	 		//c.addDefault("KeepSpawnChunksInMemory", false);
   	 		c.options().copyDefaults(true);
   		    m.saveConfig();
   		    
			m.log("ES: Adding Header to config.yml... ");
			StringBuffer contents = new StringBuffer();
			try {
				BufferedReader reader = new BufferedReader(new InputStreamReader(m.getClass().getResourceAsStream("configheader.txt")));
				String text = new String();
				while ((text = reader.readLine()) != null) {
					contents.append(text).append(System.getProperty("line.separator"));
				}
				try {
					if (reader != null) {
						reader.close();
					}
				} catch (IOException e) {
					e.printStackTrace();
				}
			} catch (FileNotFoundException e) {
				e.printStackTrace();
			} catch (IOException e) {
				e.printStackTrace();
			} finally {
				c.options().header(contents.toString());
				c.options().copyHeader(true);
				m.saveConfig();
			}
		}
	}
	
	public void odCheck() {
		final FileConfiguration c = m.getConfig();
		if (c.contains("UseSpawnFlags") || c.contains("LimitSpawners")) {
  			m.log("ES ALERT: Outdated Config! Please refer to");
			m.log("http://dev.bukkit.org/server-mods/entitysuppressor/pages/main/configuration/");
			m.log("or re-generate your config for...");
			m.log("...information about updating your config.");
  		}
	}
	
	public void load() {
		final FileConfiguration c = m.getConfig();
  		m.log("ES: Loading Config File... ");
  		odCheck();
  		lMonsters = c.getBoolean("LimitMonsters", false);
  		lSquid = c.getBoolean("LimitSquid", false);
  		lAnimals = c.getBoolean("LimitAnimals", false);
  		lBats = c.getBoolean("LimitBats", false);
  		lNPCs = c.getBoolean("LimitNPCs", false);
  		interval = c.getInt("ScanInterval", 100);
  		if (interval < 20) {
  			interval = 20;
  			m.log("ES NOTICE: Interval time is too low! Using 20 ticks instead.");
  		}
		entitiesPerScan = c.getInt("EntitiesPerScan", 64);
  		
  		loadESWByKey("MonsterMaximums", c);
  		loadESWByKey("SquidMaximums", c);
  		loadESWByKey("AnimalMaximums", c);
  		loadESWByKey("BatMaximums", c);
  		loadESWByKey("NPCMaximums", c);
  		loadESWByKey("ChunkCalculatedMonsterMaximums", c);
  		loadESWByKey("ChunkCalculatedSquidMaximums", c);
  		loadESWByKey("ChunkCalculatedAnimalMaximums", c);
  		loadESWByKey("ChunkCalculatedBatMaximums", c);
  		loadESWByKey("ChunkCalculatedNPCMaximums", c);
  		
		uRemoveM = c.getBoolean("UseMonsterDistanceRemoval", true);
		sqRemovalDist = (int)Math.pow(c.getInt("RemoveAtDistance", 128), 2);
		sqCancelDist = (int)Math.pow(c.getInt("CancelSpawnAtDistance", 92), 2);
		uLRemove = c.getBoolean("OnlySpawnOnCloseLayers", true);
		maxLDev = c.getInt("MaximumSpawnLayerDeviation", 20);
		ticksPerMonsterSpawns = c.getInt("TicksPerMonsterSpawns", 10);
		if (ticksPerMonsterSpawns < 5) {
			ticksPerMonsterSpawns = 5;
			m.log("ES NOTICE: to few ticks between monster spawns! Using 5 ticks instead.");
		}
		else if (ticksPerMonsterSpawns > 20) {
			ticksPerMonsterSpawns = 20;
			m.log("ES NOTICE: to many ticks between monster spawns! Using 20 ticks instead.");
		}
		ticksPerAnimalSpawns = c.getInt("TicksPerAnimalSpawns", 800);
		if (ticksPerAnimalSpawns < 400) {
			ticksPerAnimalSpawns = 400;
			m.log("ES NOTICE: to few ticks between animal spawns! Using 400 ticks instead.");
		}
		else if (ticksPerAnimalSpawns > 1600) {
			ticksPerAnimalSpawns = 1600;
			m.log("ES NOTICE: to many ticks between animal spawns! Using 1600 ticks instead.");
		}
		qSpawner = c.getBoolean("QueueSpawnerSpawns", true);
		KSIM = c.contains("KeepSpawnChunksInMemory");
		keepSpawnInMem = c.getBoolean("KeepSpawnChunksInMemory", false);
		d = c.getBoolean("Debug", false);
		if (d) m.log("ES NOTICE: Running in debug mode!");
  	}
	
	private void loadESWByKey(String key, FileConfiguration c) {
  		if (c.contains(key)) {
  			for (Entry<String, Object> entry : c.getConfigurationSection(key).getValues(true).entrySet()) {
  				if (!Utils.eswLookup.containsKey(entry.getKey())) {
  					Utils.eswLookup.put(entry.getKey(), new ESWorld(entry.getKey()));
  				}
  				Utils.eswLookup.get(entry.getKey()).setMaxByString(key, Integer.parseInt(entry.getValue().toString().trim()));
  			}
  		}
	}
}
